File: src/main/java/me/rohank05/echo/EchoConverter.java
package me.rohank05.echo;

import com.github.natanbc.lavadsp.Converter;

public class EchoConverter implements Converter {
    private final int sampleRate;
    private float[] echoBuffer;
    private double delay;
    private int position;
    private float decay;
    public EchoConverter(int sampleRate){
        this.sampleRate = sampleRate;
    }

    /**
     * @param delay A new echo buffer delay in seconds.
     */
    public void setDelay(double delay) {
        this.delay = delay;
    }

    /**
     * A decay, should be a value between zero and one.
     * @param decay the new decay (preferably between zero and one).
     */
    public void setDecay(float decay) {
        this.decay = decay;
    }

    //Apply Echo length to the buffer
    private void applyNewEchoLength(){
        if(this.delay == -1) return;
        float[] newEchoBuffer = new float[(int) (sampleRate * delay)];
        if(echoBuffer != null){
            for(int i = 0; i< newEchoBuffer.length; i++){
                if(position >= echoBuffer.length){
                    position = 0;
                }
                newEchoBuffer[i] = echoBuffer[position];
                position++;
            }
        }
        this.echoBuffer = newEchoBuffer;
        delay = -1;

    }

    @Override
    public void process(float[] input, int inputOffset, float[] output, int outputOffset, int samples) {
        if(echoBuffer != null){
            for(int i = 0; i < samples; i++){
                if(position >= echoBuffer.length){
                    position = 0;
                }
                float current = input[i+inputOffset];
                //output is the input added with the decayed echo
                current = current + echoBuffer[position] * decay;
                //store the sample in the buffer
                echoBuffer[position] = current;
                position++;
                output[i+outputOffset] = current;
            }
        }
        applyNewEchoLength();
    }
}

File: src/main/kotlin/com/therohankumar/echo/EchoConverter.kt
package com.therohankumar.echo

import com.github.natanbc.lavadsp.Converter
import kotlin.math.min

class EchoConverter(private val sampleRate: Int) : Converter {
    private var buffer: FloatArray = FloatArray(0)
    private var position = 0
    private var decay = 0.5f
    private var pendingDelay = -1.0

    fun setDelay(seconds: Double) {
        require(seconds > 0) { "Delay must be positive" }
        pendingDelay = seconds
    }

    fun setDecay(value: Float) {
        require(value in 0f..1f) { "Decay must be between 0 and 1" }
        decay = value
    }

    private fun resizeBuffer() {
        if (pendingDelay == -1.0) return

        val newBuffer = FloatArray((sampleRate * pendingDelay).toInt())
        if (buffer.isNotEmpty()) {
            val copyLength = min(buffer.size, newBuffer.size)
            System.arraycopy(buffer, position, newBuffer, 0, copyLength - position)
            System.arraycopy(buffer, 0, newBuffer, copyLength - position, position)
        }

        buffer = newBuffer
        position = 0
        pendingDelay = -1.0
    }

    override fun process(input: FloatArray, inputOffset: Int, output: FloatArray, outputOffset: Int, samples: Int) {
        if (buffer.isEmpty()) {
            input.copyInto(output, outputOffset, inputOffset, inputOffset + samples)
            resizeBuffer()
            return
        }

        for (i in 0 until samples) {
            val current = input[i + inputOffset] + buffer[position] * decay
            buffer[position] = current
            output[i + outputOffset] = current

            position = (position + 1) % buffer.size
        }

        resizeBuffer()
    }
}